src/Streamly/Internal/Data/Parser/Tee.hs

```diff
@@ -0,0 +1,297 @@
+{-# LANGUAGE CPP                       #-}
+
+-- Somehow uni-pattern match using "let" produces better optimized code
+-- compared to using cases and using explicit "error" messages in unreachable
+-- cases.  There seem to be no way to silence individual warning so we use this
+-- global option for the file.
+--
+-- Disabling this has the potential to mask off some legit warnings, therefore
+-- we have segregated code that uses uni-pattern matches in this file.
+#if __GLASGOW_HASKELL__ >= 800
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+#endif
+
+#include "inline.hs"
+
+-- |
+-- Module      : Streamly.Internal.Data.Parser.Tee
+-- Copyright   : (c) 2020 Composewell Technologies
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- Parsers.
+
+module Streamly.Internal.Data.Parser.Tee
+    (
+    -- Parallel parsers
+      teeWith
+    , teeWithFst
+    , teeWithMin
+    )
+where
+
+import Control.Exception (assert)
+import Prelude
+       hiding (any, all, takeWhile)
+
+import Fusion.Plugin.Types (Fuse(..))
+import Streamly.Internal.Data.Parser.Types (Parser(..), Step(..))
+
+-------------------------------------------------------------------------------
+-- Distributing Parsers
+-------------------------------------------------------------------------------
+
+data StepState s a = StepState s | StepResult a
+
+-- | State of the pair of parsers in a tee composition
+-- Note: strictness annotation is important for fusing the constructors
+{-# ANN type TeeState Fuse #-}
+data TeeState sL sR x a b =
+    TeePair !([x], StepState sL a, [x], [x]) !([x], StepState sR b, [x], [x])
+
+data Res = Yld Int | Stp Int | Skp
+
+-- XXX: With the current "Step" semantics, it is hard to write, and not sure
+-- how useful, an efficient teeWith that returns a correct unused input count.
+--
+-- | @teeWith f p1 p2@ distributes its input to both @p1@ and @p2@ until both
+-- of them succeed or fail and combines their output using @f@. The parser
+-- succeeds if both the parsers succeed.
+--
+-- /Internal/
+--
```

Comment:
Just a reminded that this function's implementation is complicated, we shouldn't expose this combinator from non-internal modules without tests.